proto/merge.go

```diff
@@ -48,64 +49,101 @@ func mergedFileDescriptors(debug bool) (*descriptorpb.FileDescriptorSet, error)
+				}
+				return
+			}
 
-		fd := &descriptorpb.FileDescriptorProto{}
-		err = protov2.Unmarshal(bz, fd)
-		if err != nil {
-			return nil, err
-		}
+			fd := &descriptorpb.FileDescriptorProto{}
+			err = protov2.Unmarshal(bz, fd)
+			if err != nil {
+				chFd <- &fdResult{
+					Err: err,
+				}
+				return
+			}
 
-		err = CheckImportPath(fd.GetName(), fd.GetPackage())
-		if err != nil {
-			checkImportErr = append(checkImportErr, err.Error())
-		}
+			err = CheckImportPath(fd.GetName(), fd.GetPackage())
+			if err != nil {
+				checkImportErr = append(checkImportErr, err.Error())
+			}
 
-		// If it's not in the protoregistry file descriptors, add it.
-		protoregFd, err := protoregistry.GlobalFiles.FindFileByPath(*fd.Name)
-		// If we already loaded gogo's file descriptor, compare that the 2
-		// are strictly equal, or else log a warning.
-		if err != nil {
-			// If we have a NotFound error, we add this file descriptor to the
-			// final file descriptor set.
-			if errors.Is(err, protoregistry.NotFound) {
-				fds.File = append(fds.File, fd)
+			// If it's not in the protoregistry file descriptors, add it.
+			protoregFd, err := protoregistry.GlobalFiles.FindFileByPath(*fd.Name)
+			// If we already loaded gogo's file descriptor, compare that the 2
+			// are strictly equal, or else log a warning.
+			if err != nil {
+				// If we have a NotFound error, we add this file descriptor to the
+				// final file descriptor set.
+				if errors.Is(err, protoregistry.NotFound) {
+					chFd <- &fdResult{
+						Fd: fd,
+					}
+				} else {
+					chFd <- &fdResult{
+						Err: err,
+					}
+					return
+				}
 			} else {
-				return nil, err
-			}
-		} else {
-			// If there's a mismatch, we log a warning. If there was no
-			// mismatch, then we do nothing, and take the protoregistry file
-			// descriptor as the correct one.
-			if !protov2.Equal(protodesc.ToFileDescriptorProto(protoregFd), fd) {
-				diff := cmp.Diff(protodesc.ToFileDescriptorProto(protoregFd), fd, protocmp.Transform())
-				diffErr = append(diffErr, fmt.Sprintf("Mismatch in %s:\n%s", *fd.Name, diff))
+				// If there's a mismatch, we log a warning. If there was no
+				// mismatch, then we do nothing, and take the protoregistry file
+				// descriptor as the correct one.
+				if debug && !protov2.Equal(protodesc.ToFileDescriptorProto(protoregFd), fd) {
+					diff := cmp.Diff(protodesc.ToFileDescriptorProto(protoregFd), fd, protocmp.Transform())
+					diffErr = append(diffErr, fmt.Sprintf("Mismatch in %s:\n%s", *fd.Name, diff))
+				}
 			}
+		}(compressedBz)
+	}
+	go func() {
+		wg.Wait()
+		close(chFd)
+	}()
+
+	haveFileDescriptor := map[string]bool{}
+	for r := range chFd {
+		if r.Err != nil {
+			return nil, r.Err
 		}
+		fds.File = append(fds.File, r.Fd)
+		haveFileDescriptor[*r.Fd.Name] = true
 	}
+	// Add protoregistry file descriptors to our final file descriptor set.
+	protoregistry.GlobalFiles.RangeFiles(func(fileDescriptor protoreflect.FileDescriptor) bool {
```

Comment:
If we can build an array of these file descriptors first, we could process them in parallel too

Reply:
Not sure if I misunderstand, but not much speed gain when I change [RangeFiles](https://github.com/protocolbuffers/protobuf-go/blob/v1.29.1/reflect/protoregistry/registry.go#L350) to
```
func (r *Files) RangeFiles() <-chan protoreflect.FileDescriptor {
	ch := make(chan protoreflect.FileDescriptor)
	if r == nil {
		close(ch)
		return ch
	}
	if r == GlobalFiles {
		globalMutex.RLock()
		defer globalMutex.RUnlock()
	}
	go func() {
		defer close(ch)
		var wg sync.WaitGroup
		for _, files := range r.filesByPath {
			for _, file := range files {
				wg.Add(1)
				go func(file protoreflect.FileDescriptor) {
					defer wg.Done()
					select {
					case ch <- file:
					default:
						return
					}
				}(file)
			}
		}
		wg.Wait()
	}()
	return ch
}
```